From 20ca708285f75d8a77e675bbbeda61bde5653259 Mon Sep 17 00:00:00 2001
From: antheas <git@antheas.dev>
Date: Sat, 22 Jun 2024 01:44:50 +0200
Subject: [PATCH 1/3] place all /etc files in the last layer to avoid lib32
 errors

---
 rust/src/container.rs | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/rust/src/container.rs b/rust/src/container.rs
index 5a14c7b0..3d1c93cf 100644
--- a/rust/src/container.rs
+++ b/rust/src/container.rs
@@ -391,6 +391,17 @@ pub fn container_encapsulate(args: Vec<String>) -> CxxResult<()> {
         for (nevra, pkgmeta) in package_meta.iter() {
             for path in pkgmeta.provided_paths()? {
                 let path = Utf8PathBuf::from(path);
+                if path.starts_with("/etc") || path.starts_with("/usr/etc") {
+                    // Place all /etc files to the last layer.
+                    // lib32 packages can contain the same etc file, 
+                    // which causes ostree-rs-ext to emit the same etc file
+                    // twice and corrupt the tar archive.
+                    // In addition, this pulls /etc modifications to the image
+                    // layer, which would otherwise invalidate the cache on changes.
+                    // For a large image, they are around 40mb uncompressed
+                    // and negligible compressed.
+                    continue;
+                }
 
                 // Resolve the path to its ostree file
                 if let Some(ostree_paths) = fsutil::resolve_ostree_paths(
-- 
2.45.2

