From 5c2c5fc81a9b0c0d29467ce07bee668776595e2b Mon Sep 17 00:00:00 2001
From: antheas <git@antheas.dev>
Date: Sat, 22 Jun 2024 01:45:17 +0200
Subject: [PATCH 2/3] switch to using warning for packages with the same name

---
 src/libpriv/rpmostree-refts.cxx | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/libpriv/rpmostree-refts.cxx b/src/libpriv/rpmostree-refts.cxx
index 46f63621..8ad003dd 100644
--- a/src/libpriv/rpmostree-refts.cxx
+++ b/src/libpriv/rpmostree-refts.cxx
@@ -165,13 +165,15 @@ RpmTs::package_meta (const rust::Str name) const
     {
       // TODO: Somehow we get two `libgcc-8.5.0-10.el8.x86_64` in current RHCOS, I don't
       // understand that.
+      // lib32 packages will have the same name. Steam is a lib32 package.
+      // Lutris also pulls in wine lib32 packages, one of which is NetworkManager-libnm.
+      // This means that immutable gaming images will have packages with
+      // the same name and would error out. Emit a warning and continue.
       if (retval != nullptr)
         {
           auto nevra = header_get_nevra (h);
-          g_autofree char *buf
-              = g_strdup_printf ("Multiple installed '%s' (%s, %s)", name_c.c_str (),
-                                 retval->nevra ().c_str (), nevra.c_str ());
-          throw std::runtime_error (buf);
+          g_print ("Warning! Multiple installed '%s' (%s, %s).\n",
+                   name_c.c_str (), retval->nevra ().c_str (), nevra.c_str ());
         }
 
       retval = std::make_unique<PackageMeta> (h);
-- 
2.45.2

